R/datasummary_crosstab.R

```diff
@@ -0,0 +1,58 @@
+#' Cross tabulations for categorical variables
+#'
+#' Convenience function to tabulate counts, cell percentages, and row/column
+#' percentages for categorical variables.
+#'
+#' @param formula A two-sided formula to describe the table: rows ~ columns, where
+#' rows and columns are variables in the data. Both rows and column may contain
+#' interactions, e.g., var1 * var2 ~ var3.
+#' @param statistic A formula of the form 1 ~ 1 + N + Percent("row"). The left-hand
+#' side may only be empty or contain a 1 to include row totals.
+#' The right-hand side may contain: 1 for column totals, N for counts, Percent() for
+#' cell percentages, Percent("row") for row percentages, Percent("col") for column
+#' percentages.
+#' @param data A data.frame (or tibble)
+#' @param sparse_header TRUE or FALSE. TRUE eliminates column headers which
+#' have a unique label across all columns, except for the row immediately above
+#' the data. FALSE keeps all headers. The order in which terms are entered in
+#' the formula determines the order in which headers appear.
+#' @param ... passed to datasummary
+#' @return a crosstab
+#' @details
+#' X
+#' @examples
+#' \dontrun{
+#' }
+#'
+#' @export
+datasummary_crosstab <- function(
+        formula,
+        statistic = 1 ~ 1 + N + Percent("row"),
+        data,
+        sparse_header = FALSE, ...) {
+    checkmate::assert_formula(formula)
+    checkmate::assert_formula(statistic)
+    checkmate::assert_data_frame(data, min.rows = 1, min.cols = 1)
+
+    lhs <- paste0("Factor(", all.vars(formula[[2]]), ")")
+    rhs <- paste0("Factor(", all.vars(formula[[3]]), ")")
+
+    # find out if row/column totals should be included
+    terms <- stats::terms(statistic)
+    total_row <- ifelse(attr(terms, "response") == 1, " + 1", "")
+    indep <- tail(as.character(statistic), 1)
+    indep_vars <- unlist(strsplit(indep, "+", fixed = TRUE))
+    total_col <- ifelse("1" %in% trimws(indep_vars), " + 1", "")
+
+    # adjust labels for %
+    labels <- labels(terms)
+    labels[labels == 'Percent()'] <- 'Heading("%")*Percent()'
+    labels[labels == 'Percent("row")'] <- 'Heading("% row")*Percent("row")'
+    labels[labels == 'Percent("col")'] <- 'Heading("% col")*Percent("col")'
+
+    d_formula <- sprintf("(%s%s) * (%s) ~ %s%s",
```

Comment:
What happens if users do: `var1 + var2 ~ var3`? I guess this overrides with `*` without warning, which is probably not desirable. Would it be difficult to allow the plus sign, which would give something like two rbind-ed cross-tabs. Is that a rabbit hole?

Reply:
Might be a rabbit hole, but I'll see if it's difficult. Currently, the plan was to just raise an error when users chain variables with +.

Reply:
Cool. I'm really terrible at formula manipulation, but what would be really cool is if we could find an answer to the StackOverflow question I just posted and use that instead: https://stackoverflow.com/questions/67434078/r-formula-wrap-all-variables-in-a-transformation

Reply:
Looks like I got an answer that might work for us. Posting it in the main thread.